Firstly I load the packages I need.

In [2]:
import io
import requests
import pandas as pd
import numpy as np

Secondly I use the pandas .read_html method to scrape the table with the neighbourhoods in Toronto from the Wikipedia page. I specified by using [0] for the particular table I want since the page contains more than one table. 

In [3]:
Toronto = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header=0,
                     )
Toronto[0]

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
..       ...               ...   
257      M9X      Not assigned   
258      M1Y      Not assigned   
259      M2Y      Not assigned   
260      M3Y      Not assigned   
261      M4Y  Downtown Toronto   
262      M5Y      Not assigned   
263      M6Y      Not assigned   
264      M7Y      East Toronto   
265      M8Y         Etobicoke   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
273      M9Y      Not assigned   
274      M1Z      Not assigned   
275      M2Z      Not assigned   
276      M3Z      Not assigned   
277      M4Z      Not assigned   
278      M5Z      Not assigned   
279      M6Z      Not assigned   
280      M7Z      Not assigned   
281      M8Z         Etobicoke   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   
286      M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                     Lawrence Heights  
6                                       Lawrence Manor  
7                                         Not assigned  
8                                         Not assigned  
9                                         Queen's Park  
10                                               Rouge  
11                                             Malvern  
12                                        Not assigned  
13                                     Don Mills North  
14                                    Woodbine Gardens  
15                                       Parkview Hill  
16                                             Ryerson  
17                                     Garden District  
18                                           Glencairn  
19                                        Not assigned  
20                                        Not assigned  
21                                          Cloverdale  
22                                           Islington  
23                                        Martin Grove  
24                                    Princess Gardens  
25                                     West Deane Park  
26                                      Highland Creek  
27                                          Rouge Hill  
28                                          Port Union  
29                                        Not assigned  
..                                                 ...  
257                                       Not assigned  
258           

The scraped table contains 287 rows and 3 columns. Then I convert the table into a pandas dataframe called df.

In [4]:
df = pd.DataFrame(Toronto[0])
df

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
..       ...               ...   
257      M9X      Not assigned   
258      M1Y      Not assigned   
259      M2Y      Not assigned   
260      M3Y      Not assigned   
261      M4Y  Downtown Toronto   
262      M5Y      Not assigned   
263      M6Y      Not assigned   
264      M7Y      East Toronto   
265      M8Y         Etobicoke   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
273      M9Y      Not assigned   
274      M1Z      Not assigned   
275      M2Z      Not assigned   
276      M3Z      Not assigned   
277      M4Z      Not assigned   
278      M5Z      Not assigned   
279      M6Z      Not assigned   
280      M7Z      Not assigned   
281      M8Z         Etobicoke   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   
286      M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                     Lawrence Heights  
6                                       Lawrence Manor  
7                                         Not assigned  
8                                         Not assigned  
9                                         Queen's Park  
10                                               Rouge  
11                                             Malvern  
12                                        Not assigned  
13                                     Don Mills North  
14                                    Woodbine Gardens  
15                                       Parkview Hill  
16                                             Ryerson  
17                                     Garden District  
18                                           Glencairn  
19                                        Not assigned  
20                                        Not assigned  
21                                          Cloverdale  
22                                           Islington  
23                                        Martin Grove  
24                                    Princess Gardens  
25                                     West Deane Park  
26                                      Highland Creek  
27                                          Rouge Hill  
28                                          Port Union  
29                                        Not assigned  
..                                                 ...  
257                                       Not assigned  
258           

I then delete all cells with a Borough that is Not assigned so that I only process the cells that have an assigned borough.

In [5]:
df = df[df.Borough != 'Not assigned']
df

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
..       ...               ...   
234      M9V         Etobicoke   
235      M1W       Scarborough   
238      M4W  Downtown Toronto   
239      M5W  Downtown Toronto   
242      M8W         Etobicoke   
243      M8W         Etobicoke   
244      M9W         Etobicoke   
245      M1X       Scarborough   
248      M4X  Downtown Toronto   
249      M4X  Downtown Toronto   
250      M5X  Downtown Toronto   
251      M5X  Downtown Toronto   
254      M8X         Etobicoke   
255      M8X         Etobicoke   
256      M8X         Etobicoke   
261      M4Y  Downtown Toronto   
264      M7Y      East Toronto   
265      M8Y         Etobicoke   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
281      M8Z         Etobicoke   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                     Lawrence Heights  
6                                       Lawrence Manor  
7                                         Not assigned  
9                                         Queen's Park  
10                                               Rouge  
11                                             Malvern  
13                                     Don Mills North  
14                                    Woodbine Gardens  
15                                       Parkview Hill  
16                                             Ryerson  
17                                     Garden District  
18                                           Glencairn  
21                                          Cloverdale  
22                                           Islington  
23                                        Martin Grove  
24                                    Princess Gardens  
25                                     West Deane Park  
26                                      Highland Creek  
27                                          Rouge Hill  
28                                          Port Union  
30                                     Flemingdon Park  
31                                     Don Mills South  
32                                    Woodbine Heights  
33                                      St. James Town  
34                                  Humewood-Cedarvale  
37                                   Bloordale Gardens  
38                                            Eringate  
..                                                 ...  
234                                        Thistletown  
235           

This reduces my dataframe from 287 rows to 210 rows, meaning that 77 cells had a Borough that was Not assigned. 

I then combine into one row all Neighbourhoods with the same Postcode, with the Neighbourhoods separated with a comma. To do that, firstly I use the .groupby method on condition of the Neighbourhoods having the same Postcode and then use the .apply method to execute the grouping and the .join method to separate the Neighbourhoods with a coma and save the result in a new dataframe called df1.

In [6]:
df1 = df.groupby('Postcode')['Neighbourhood'].apply(','.join).reset_index()
df1

Postcode                                      Neighbourhood
0        M1B                                      Rouge,Malvern
1        M1C               Highland Creek,Rouge Hill,Port Union
2        M1E                    Guildwood,Morningside,West Hill
3        M1G                                             Woburn
4        M1H                                          Cedarbrae
5        M1J                                Scarborough Village
6        M1K          East Birchmount Park,Ionview,Kennedy Park
7        M1L                      Clairlea,Golden Mile,Oakridge
8        M1M      Cliffcrest,Cliffside,Scarborough Village West
9        M1N                         Birch Cliff,Cliffside West
10       M1P  Dorset Park,Scarborough Town Centre,Wexford He...
11       M1R                                   Maryvale,Wexford
12       M1S                                          Agincourt
13       M1T              Clarks Corners,Sullivan,Tam O'Shanter
14       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...
15       M1W                                    L'Amoreaux West
16       M1X                                        Upper Rouge
17       M2H                                  Hillcrest Village
18       M2J                         Fairview,Henry Farm,Oriole
19       M2K                                    Bayview Village
20       M2L                            Silver Hills,York Mills
21       M2M                             Newtonbrook,Willowdale
22       M2N                                   Willowdale South
23       M2P                                    York Mills West
24       M2R                                    Willowdale West
25       M3A                                          Parkwoods
26       M3B                                    Don Mills North
27       M3C                    Flemingdon Park,Don Mills South
28       M3H      Bathurst Manor,Downsview North,Wilson Heights
29       M3J                     Northwood Park,York University
..       ...                                                ...
73       M6C                                 Humewood-Cedarvale
74       M6E                                Caledonia-Fairbanks
75       M6G                                           Christie
76       M6H                        Dovercourt Village,Dufferin
77       M6J                            Little Portugal,Trinity
78       M6K         Brockton,Exhibition Place,Parkdale Village
79       M6L                   Downsview,North Park,Upwood Park
80       M6M        Del Ray,Keelesdale,Mount Dennis,Silverthorn
81       M6N                       The Junction North,Runnymede
82       M6P                       High Park,The Junction South
83       M6R                              Parkdale,Roncesvalles
84       M6S                                  Runnymede,Swansea
85       M7A                                       Not assigned
86       M7R              Canada Post Gateway Processing Centre
87       M7Y  Business Reply Mail Processing Centre 969 Eastern
88       M8V         Humber Bay Shores,Mimico South,New Toronto
89       M8W                              Alderwood,Long Branch
90       M8X        The Kingsway,Montgomery Road,Old Mill North
91       M8Y  Humber Bay,King's Mill Park,Kingsway Park Sout...
92       M8Z  Kingsway Park South West,Mimico NW,The Queensw...
93       M9A                                       Queen's Park
94       M9B  Cloverdale,Islington,Martin Grove,Princess Gar...
95       M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...
96       M9L                                      Humber Summit
97       M9M                                    Emery,Humberlea
98       M9N                                             Weston
99       M9P                                          Westmount
100      M9R  Kingsview Village,Martin Grove Gardens,Richvie...
101      M9V  Albion Gardens,Beaumond Heights,Humbergate,Jam...
102      M9W                                          Northwest

[103 rows x 2 columns]

The returned output above excludes the Borough column. I use the pandas .merge method to combine the Borough column from the df dataframe and the df1 dataframe on condition that they're merged on matching the Borough to the Neighbourhood based on the Postcode and I save the results in a new dataframe called df2.

In [7]:
df2 = pd.merge(df1, df[["Postcode", "Borough"]], on="Postcode")
df2

Postcode                                      Neighbourhood       Borough
0        M1B                                      Rouge,Malvern   Scarborough
1        M1B                                      Rouge,Malvern   Scarborough
2        M1C               Highland Creek,Rouge Hill,Port Union   Scarborough
3        M1C               Highland Creek,Rouge Hill,Port Union   Scarborough
4        M1C               Highland Creek,Rouge Hill,Port Union   Scarborough
5        M1E                    Guildwood,Morningside,West Hill   Scarborough
6        M1E                    Guildwood,Morningside,West Hill   Scarborough
7        M1E                    Guildwood,Morningside,West Hill   Scarborough
8        M1G                                             Woburn   Scarborough
9        M1H                                          Cedarbrae   Scarborough
10       M1J                                Scarborough Village   Scarborough
11       M1K          East Birchmount Park,Ionview,Kennedy Park   Scarborough
12       M1K          East Birchmount Park,Ionview,Kennedy Park   Scarborough
13       M1K          East Birchmount Park,Ionview,Kennedy Park   Scarborough
14       M1L                      Clairlea,Golden Mile,Oakridge   Scarborough
15       M1L                      Clairlea,Golden Mile,Oakridge   Scarborough
16       M1L                      Clairlea,Golden Mile,Oakridge   Scarborough
17       M1M      Cliffcrest,Cliffside,Scarborough Village West   Scarborough
18       M1M      Cliffcrest,Cliffside,Scarborough Village West   Scarborough
19       M1M      Cliffcrest,Cliffside,Scarborough Village West   Scarborough
20       M1N                         Birch Cliff,Cliffside West   Scarborough
21       M1N                         Birch Cliff,Cliffside West   Scarborough
22       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   Scarborough
23       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   Scarborough
24       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   Scarborough
25       M1R                                   Maryvale,Wexford   Scarborough
26       M1R                                   Maryvale,Wexford   Scarborough
27       M1S                                          Agincourt   Scarborough
28       M1T              Clarks Corners,Sullivan,Tam O'Shanter   Scarborough
29       M1T              Clarks Corners,Sullivan,Tam O'Shanter   Scarborough
..       ...                                                ...           ...
180      M8Z  Kingsway Park South West,Mimico NW,The Queensw...     Etobicoke
181      M8Z  Kingsway Park South West,Mimico NW,The Queensw...     Etobicoke
182      M9A                                       Queen's Park  Queen's Park
183      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
184      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
185      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
186      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
187      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
188      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
189      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
190      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
191      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
192      M9L                                      Humber Summit    North York
193      M9M                                    Emery,Humberlea    North York
194      M9M                                    Emery,Humberlea    North York
195      M9N                                             Weston          York
196      M9P                                          Westmount     Etobicoke
197      M9R  Kingsview Village,Martin Grove Gardens,Richvie...     Etobicoke
198      M9R  Kingsview Village,Martin Grove Gardens,Richvie...     Etobicoke
199      M9R  Kingsview Vi

I then rearrange the order of my columns to start with Postcode, followed by Borough and thirdly Neighbourhood.

In [8]:
df2 = df2[['Postcode', 'Borough', 'Neighbourhood']]
df2

Postcode       Borough                                      Neighbourhood
0        M1B   Scarborough                                      Rouge,Malvern
1        M1B   Scarborough                                      Rouge,Malvern
2        M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
3        M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
4        M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
5        M1E   Scarborough                    Guildwood,Morningside,West Hill
6        M1E   Scarborough                    Guildwood,Morningside,West Hill
7        M1E   Scarborough                    Guildwood,Morningside,West Hill
8        M1G   Scarborough                                             Woburn
9        M1H   Scarborough                                          Cedarbrae
10       M1J   Scarborough                                Scarborough Village
11       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
12       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
13       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
14       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
15       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
16       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
17       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
18       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
19       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
20       M1N   Scarborough                         Birch Cliff,Cliffside West
21       M1N   Scarborough                         Birch Cliff,Cliffside West
22       M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
23       M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
24       M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
25       M1R   Scarborough                                   Maryvale,Wexford
26       M1R   Scarborough                                   Maryvale,Wexford
27       M1S   Scarborough                                          Agincourt
28       M1T   Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
29       M1T   Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
..       ...           ...                                                ...
180      M8Z     Etobicoke  Kingsway Park South West,Mimico NW,The Queensw...
181      M8Z     Etobicoke  Kingsway Park South West,Mimico NW,The Queensw...
182      M9A  Queen's Park                                       Queen's Park
183      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
184      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
185      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
186      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
187      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
188      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
189      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
190      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
191      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
192      M9L    North York                                      Humber Summit
193      M9M    North York                                    Emery,Humberlea
194      M9M    North York                                    Emery,Humberlea
195      M9N          York                                             Weston
196      M9P     Etobicoke                                          Westmount
197      M9R     Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
198      M9R     Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
199      M9R     Etobicoke

I then drop all resulting duplicates based on the same Postcode and save the results in a new dataframe called df3.

In [9]:
df3 = df2.drop_duplicates(subset="Postcode")
df3

Postcode           Borough  \
0        M1B       Scarborough   
2        M1C       Scarborough   
5        M1E       Scarborough   
8        M1G       Scarborough   
9        M1H       Scarborough   
10       M1J       Scarborough   
11       M1K       Scarborough   
14       M1L       Scarborough   
17       M1M       Scarborough   
20       M1N       Scarborough   
22       M1P       Scarborough   
25       M1R       Scarborough   
27       M1S       Scarborough   
28       M1T       Scarborough   
31       M1V       Scarborough   
35       M1W       Scarborough   
36       M1X       Scarborough   
37       M2H        North York   
38       M2J        North York   
41       M2K        North York   
42       M2L        North York   
44       M2M        North York   
46       M2N        North York   
47       M2P        North York   
48       M2R        North York   
49       M3A        North York   
50       M3B        North York   
51       M3C        North York   
53       M3H        North York   
56       M3J        North York   
..       ...               ...   
133      M6C              York   
134      M6E              York   
135      M6G  Downtown Toronto   
136      M6H      West Toronto   
138      M6J      West Toronto   
140      M6K      West Toronto   
143      M6L        North York   
146      M6M              York   
150      M6N              York   
152      M6P      West Toronto   
154      M6R      West Toronto   
156      M6S      West Toronto   
158      M7A      Queen's Park   
159      M7R       Mississauga   
160      M7Y      East Toronto   
161      M8V         Etobicoke   
164      M8W         Etobicoke   
166      M8X         Etobicoke   
169      M8Y         Etobicoke   
177      M8Z         Etobicoke   
182      M9A      Queen's Park   
183      M9B         Etobicoke   
188      M9C         Etobicoke   
192      M9L        North York   
193      M9M        North York   
195      M9N              York   
196      M9P         Etobicoke   
197      M9R         Etobicoke   
201      M9V         Etobicoke   
209      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
2                 Highland Creek,Rouge Hill,Port Union  
5                      Guildwood,Morningside,West Hill  
8                                               Woburn  
9                                            Cedarbrae  
10                                 Scarborough Village  
11           East Birchmount Park,Ionview,Kennedy Park  
14                       Clairlea,Golden Mile,Oakridge  
17       Cliffcrest,Cliffside,Scarborough Village West  
20                          Birch Cliff,Cliffside West  
22   Dorset Park,Scarborough Town Centre,Wexford He...  
25                                    Maryvale,Wexford  
27                                           Agincourt  
28               Clarks Corners,Sullivan,Tam O'Shanter  
31   Agincourt North,L'Amoreaux East,Milliken,Steel...  
35                                     L'Amoreaux West  
36                                         Upper Rouge  
37                                   Hillcrest Village  
38                          Fairview,Henry Farm,Oriole  
41                                     Bayview Village  
42                             Silver Hills,York Mills  
44                              Newtonbrook,Willowdale  
46                                    Willowdale South  
47                                     York Mills West  
48                                     Willowdale West  
49                                           Parkwoods  
50                                     Don Mills North  
51                     Flemingdon Park,Don Mills South  
53       Bathurst Manor,Downsview North,Wilson Heights  
56                      Northwood Park,York University  
..                                                 ...  
133                                 Humewood-Cedarvale  
134           

The code below is to find out how many Neighbourhoods are Not assigned. The result is just one, with a Borough named Queen's Park.

In [10]:
print(df3.loc[df3['Neighbourhood'] == 'Not assigned'])

    Postcode       Borough Neighbourhood
158      M7A  Queen's Park  Not assigned


I then replace the Not assigned Neighbourhood with its Borough name by using the .loc method.

In [11]:
df3.loc[df3['Neighbourhood'] == 'Not assigned'] = "Queen's Park"
df3

/usr/local/lib/python3.5/dist-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Postcode           Borough  \
0             M1B       Scarborough   
2             M1C       Scarborough   
5             M1E       Scarborough   
8             M1G       Scarborough   
9             M1H       Scarborough   
10            M1J       Scarborough   
11            M1K       Scarborough   
14            M1L       Scarborough   
17            M1M       Scarborough   
20            M1N       Scarborough   
22            M1P       Scarborough   
25            M1R       Scarborough   
27            M1S       Scarborough   
28            M1T       Scarborough   
31            M1V       Scarborough   
35            M1W       Scarborough   
36            M1X       Scarborough   
37            M2H        North York   
38            M2J        North York   
41            M2K        North York   
42            M2L        North York   
44            M2M        North York   
46            M2N        North York   
47            M2P        North York   
48            M2R        North York   
49            M3A        North York   
50            M3B        North York   
51            M3C        North York   
53            M3H        North York   
56            M3J        North York   
..            ...               ...   
133           M6C              York   
134           M6E              York   
135           M6G  Downtown Toronto   
136           M6H      West Toronto   
138           M6J      West Toronto   
140           M6K      West Toronto   
143           M6L        North York   
146           M6M              York   
150           M6N              York   
152           M6P      West Toronto   
154           M6R      West Toronto   
156           M6S      West Toronto   
158  Queen's Park      Queen's Park   
159           M7R       Mississauga   
160           M7Y      East Toronto   
161           M8V         Etobicoke   
164           M8W         Etobicoke   
166           M8X         Etobicoke   
169           M8Y         Etobicoke   
177           M8Z         Etobicoke   
182           M9A      Queen's Park   
183           M9B         Etobicoke   
188           M9C         Etobicoke   
192           M9L        North York   
193           M9M        North York   
195           M9N              York   
196           M9P         Etobicoke   
197           M9R         Etobicoke   
201           M9V         Etobicoke   
209           M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
2                 Highland Creek,Rouge Hill,Port Union  
5                      Guildwood,Morningside,West Hill  
8                                               Woburn  
9                                            Cedarbrae  
10                                 Scarborough Village  
11           East Birchmount Park,Ionview,Kennedy Park  
14                       Clairlea,Golden Mile,Oakridge  
17       Cliffcrest,Cliffside,Scarborough Village West  
20                          Birch Cliff,Cliffside West  
22   Dorset Park,Scarborough Town Centre,Wexford He...  
25                                    Maryvale,Wexford  
27                                           Agincourt  
28               Clarks Corners,Sullivan,Tam O'Shanter  
31   Agincourt North,L'Amoreaux East,Milliken,Steel...  
35                                     L'Amoreaux West  
36                                         Upper Rouge  
37                                   Hillcrest Village  
38                          Fairview,Henry Farm,Oriole  
41                                     Bayview Village  
42                             Silver Hills,York Mills  
44                              Newtonbrook,Willowdale  
46                                    Willowdale South  
47                                     York Mills West  
48                                     Willowdale West  
49                                           Parkwoods  
50                                     Don Mills No

To ascertain the number of rows and columns in my final df3 dataframe, I use the .shape method and this shows that my final dataframe has 103 rows and 3 columns.

In [13]:
df3.shape

(103, 3)

First I use the pandas .read_csv method to read the cvs file containing the Latitudes and Longitudes of the Toronto Neighbourhoods based on Postcodes and I save the csv data in a pandas dataframe called Coords.

In [14]:
Coords = pd.read_csv("https://cocl.us/Geospatial_data") 
Coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

I then rename the Postal Code column in the Coords dataframe to Postcode like the df3 dataframe to make it easier for me to merge the 2 dataframes together.

In [15]:
Coords.rename(columns={'Postal Code':'Postcode'}, inplace=True)

print(Coords.columns)

Index(['Postcode', 'Latitude', 'Longitude'], dtype='object')


Using the pandas .merge method to combine the two dataframes (df3 and Coords) together on condition that they're merged on the Postcode being the same and I save the results in a new dataframe called df4.

In [16]:
df4 = pd.merge(df3, Coords[["Postcode", "Latitude", "Longitude"]], on="Postcode")
df4

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7R       Mississauga   
86       M7Y      East Toronto   
87       M8V         Etobicoke   
88       M8W         Etobicoke   
89       M8X         Etobicoke   
90       M8Y         Etobicoke   
91       M8Z         Etobicoke   
92       M9A      Queen's Park   
93       M9B         Etobicoke   
94       M9C         Etobicoke   
95       M9L        North York   
96       M9M        North York   
97       M9N              York   
98       M9P         Etobicoke   
99       M9R         Etobicoke   
100      M9V         Etobicoke   
101      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6            East Birchmount Park,Ionview,Kennedy Park  43.727929 -79.262029  
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577  
8        Cliffcrest,Cliffside,Scarborough Village West  43.716316 -79.239476  
9                           Birch Cliff,Cliffside West  43.692657 -79.264848  
10   Dorset Park,Scarborough Town Centre,Wexford He...  43.757410 -79.273304  
11                                    Maryvale,Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13               Clarks Corners,Sullivan,Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                          Fairview,Henry Farm,Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                             Silver Hills,York Mills  43.757490 -79.374714  
21                              Newtonbrook,Willowdale  43.789053 -79.408493  
22                                    Willowdale South  43.770120 -79.408493  

The resulting dataframe after the merge has one less row, meaning that one row was dropped maybe due to a mismatch.

## Exploring and clustering the neighbourhoods in Toronto

First I get all the dependencies that I will need, that I hadn't got as yet.

In [26]:
import sys
!{sys.executable} -m pip install geopy

     |████████████████████████████████| 102kB 17.7MB/s 


In [30]:
import sys
!{sys.executable} -m pip install folium

     |████████████████████████████████| 92kB 17.8MB/s 
     |████████████████████████████████| 133kB 34.2MB/s 
  Found existing installation: Jinja2 2.8
    Not uninstalling jinja2 at /usr/local/lib/python3.5/dist-packages, outside environment /resources/common/.virtualenv/python3
    Can't uninstall 'Jinja2'. No files were found to uninstall.


In [31]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Checking the number of Boroughs and Neighbourhoods in the dataset.

In [17]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df4['Borough'].unique()),
        df4.shape[0]
    )
)

The dataframe has 11 boroughs and 102 neighborhoods.


Firstly, in order to define an instance of the geocoder, we need to define a user_agent. I will name my user agent toronto_explorer, as shown below.

In [32]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### I then create a map of Toronto with neighborhoods superimposed on top.

In [36]:
# First I create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# then I add markers to map
for lat, lng, Borough, Neighbourhood in zip(df4['Latitude'], df4['Longitude'], df4['Borough'], df4['Neighbourhood']):
    label = '{}, {}'.format(Neighbourhood, Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

The map shows the neighbourhoods of Toronto superimposed on the map of Toronto.

I then slice the original dataframe (df4) and create a new dataframe of the neighbourhoods in North York Borough and I save the data in a new datafame called northyork_data.

In [38]:
northyork_data = df4[df4['Borough'] == 'North York'].reset_index(drop=True)
northyork_data.head()

Postcode     Borough               Neighbourhood   Latitude  Longitude
0      M2H  North York           Hillcrest Village  43.803762 -79.363452
1      M2J  North York  Fairview,Henry Farm,Oriole  43.778517 -79.346556
2      M2K  North York             Bayview Village  43.786947 -79.385975
3      M2L  North York     Silver Hills,York Mills  43.757490 -79.374714
4      M2M  North York      Newtonbrook,Willowdale  43.789053 -79.408493

Then I get the coordinates of North York.

In [39]:
address = 'North York, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York are 43.7708175, -79.4132998.


I then visualise North York with the neighbourhoods in it, as shown below.

In [41]:
# create map of North York using latitude and longitude values
map_northyork = folium.Map(location=[latitude, longitude], zoom_start=11)

# I then add markers to map
for lat, lng, label in zip(northyork_data['Latitude'], northyork_data['Longitude'], northyork_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='purple',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_northyork)  
    
map_northyork

#### Defining Foursquare Credentials and Version

In [49]:
CLIENT_ID = 'DK0DZVYAYUF2IWJNSSGAFNWQNELVW13HV2OJOMBFB04RXRIL' # your Foursquare ID
CLIENT_SECRET = '0DOANYV3WCUVAFTLL5K5WYC2CG0F2L4MQMKBRDW2V3EWTDRM' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DK0DZVYAYUF2IWJNSSGAFNWQNELVW13HV2OJOMBFB04RXRIL
CLIENT_SECRET:0DOANYV3WCUVAFTLL5K5WYC2CG0F2L4MQMKBRDW2V3EWTDRM


## 1. Exploring Neighborhoods in North York

In [52]:
radius = 500
LIMIT = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id=DK0DZVYAYUF2IWJNSSGAFNWQNELVW13HV2OJOMBFB04RXRIL&client_secret=0DOANYV3WCUVAFTLL5K5WYC2CG0F2L4MQMKBRDW2V3EWTDRM&v=20180605&ll=43.7708175,-79.4132998&radius=500&limit=30'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

 The code below runs the above function on each neighbourhood and creates a new dataframe called northyork_venues.

In [53]:
northyork_venues = getNearbyVenues(names=northyork_data['Neighbourhood'],
                                   latitudes=northyork_data['Latitude'],
                                   longitudes=northyork_data['Longitude']
                                  )



Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Bedford Park,Lawrence Manor East
Lawrence Heights,Lawrence Manor
Glencairn
Downsview,North Park,Upwood Park
Humber Summit
Emery,Humberlea


##### I then check the size of the resulting dataframe.

In [54]:
print(northyork_venues.shape)
northyork_venues.head()

(720, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Hillcrest Village               43.803762               -79.363452   
1  Hillcrest Village               43.803762               -79.363452   
2  Hillcrest Village               43.803762               -79.363452   
3  Hillcrest Village               43.803762               -79.363452   
4  Hillcrest Village               43.803762               -79.363452   

                           Venue  Venue Latitude  Venue Longitude  \
0             The Captain's Boil       43.773255       -79.413805   
1             Aroma Espresso Bar       43.769449       -79.413081   
2                        Loblaws       43.768722       -79.412101   
3  Cineplex Cinemas Empress Walk       43.768625       -79.412613   
4                  Konjiki Ramen       43.766998       -79.412222   

       Venue Category  
0  Seafood Restaurant  
1                Café  
2       Grocery Store  
3       Movie Theater  
4    Ramen Restaurant

The code below checks how many venues were returned for each neighbourhood.

In [55]:
northyork_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                           
Bathurst Manor,Downsview North,Wilson Heights                      30   
Bayview Village                                                    30   
Bedford Park,Lawrence Manor East                                   30   
CFB Toronto,Downsview East                                         30   
Don Mills North                                                    30   
Downsview Central                                                  30   
Downsview Northwest                                                30   
Downsview West                                                     30   
Downsview,North Park,Upwood Park                                   30   
Emery,Humberlea                                                    30   
Fairview,Henry Farm,Oriole                                         30   
Flemingdon Park,Don Mills South                                    30   
Glencairn                                                          30   
Hillcrest Village                                                  30   
Humber Summit                                                      30   
Lawrence Heights,Lawrence Manor                                    30   
Newtonbrook,Willowdale                                             30   
Northwood Park,York University                                     30   
Parkwoods                                                          30   
Silver Hills,York Mills                                            30   
Victoria Village                                                   30   
Willowdale South                                                   30   
Willowdale West                                                    30   
York Mills West                                                    30   

                                               Neighbourhood Longitude  Venue  \
Neighbourhood                                                                   
Bathurst Manor,Downsview North,Wilson Heights                       30     30   
Bayview Village                                                     30     30   
Bedford Park,Lawrence Manor East                                    30     30   
CFB Toronto,Downsview East                                          30     30   
Don Mills North                                                     30     30   
Downsview Central                                                   30     30   
Downsview Northwest                                                 30     30   
Downsview West                                                      30     30   
Downsview,North Park,Upwood Park                                    30     30   
Emery,Humberlea                                                     30     30   
Fairview,Henry Farm,Oriole                                          30     30   
Flemingdon Park,Don Mills South                                     30     30   
Glencairn                                                           30     30   
Hillcrest Village                                                   30     30   
Humber Summit                                                       30     30   
Lawrence Heights,Lawrence Manor                                     30     30   
Newtonbrook,Willowdale                                              30     30   
Northwood Park,York University                                      30     30   
Parkwoods                                                           30     30   
Silver Hills,York Mills                                             30     30   
Victoria Village                                                    30     30   
Willowdale South                                                    30     30   
Willowdale West                                                     30     30   
York Mills West                                                     30     30   

                                         

I want to find out how many unique categories can be curated from all the returned venues.

In [56]:
print('There are {} uniques categories.'.format(len(northyork_venues['Venue Category'].unique())))

There are 23 uniques categories.


## 2. Analyzing Each Neighbourhood

In [57]:
# one hot encoding
northyork_onehot = pd.get_dummies(northyork_venues[['Venue Category']], prefix="", prefix_sep="")

# I add neighbourhood column back to dataframe
northyork_onehot['Neighbourhood'] = northyork_venues['Neighbourhood'] 

# then I move neighbourhood column to the first column
fixed_columns = [northyork_onehot.columns[-1]] + list(northyork_onehot.columns[:-1])
northyork_onehot = northyork_onehot[fixed_columns]

northyork_onehot.head()

Neighbourhood  Arts & Crafts Store  Burrito Place  Café  Coffee Shop  \
0  Hillcrest Village                    0              0     0            0   
1  Hillcrest Village                    0              0     1            0   
2  Hillcrest Village                    0              0     0            0   
3  Hillcrest Village                    0              0     0            0   
4  Hillcrest Village                    0              0     0            0   

   Fast Food Restaurant  Grocery Store  Indonesian Restaurant  \
0                     0              0                      0   
1                     0              0                      0   
2                     0              1                      0   
3                     0              0                      0   
4                     0              0                      0   

   Japanese Restaurant  Juice Bar  Korean Restaurant  Movie Theater  \
0                    0          0                  0              0   
1                    0          0                  0              0   
2                    0          0                  0              0   
3                    0          0                  0              1   
4                    0          0                  0              0   

   Pet Store  Plaza  Ramen Restaurant  Restaurant  Sandwich Place  \
0          0      0                 0           0               0   
1          0      0                 0           0               0   
2          0      0                 0           0               0   
3          0      0                 0           0               0   
4          0      0                 1           0               0   

   Seafood Restaurant  Shopping Mall  Sports Bar  Steakhouse  Supermarket  \
0                   1              0           0           0            0   
1                   0              0           0           0            0   
2                   0              0           0           0            0   
3                   0              0           0           0            0   
4                   0              0           0           0            0   

   Sushi Restaurant  Vietnamese Restaurant  
0                 0                      0  
1                 0                      0  
2                 0                      0  
3                 0                      0  
4                 0                      0

In [58]:
# The reulting dataframe
northyork_onehot.shape

(720, 24)

###  I then group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category.

In [59]:
northyork_grouped = northyork_onehot.groupby('Neighbourhood').mean().reset_index()
northyork_grouped

Neighbourhood  Arts & Crafts Store  \
0   Bathurst Manor,Downsview North,Wilson Heights             0.033333   
1                                 Bayview Village             0.033333   
2                Bedford Park,Lawrence Manor East             0.033333   
3                      CFB Toronto,Downsview East             0.033333   
4                                 Don Mills North             0.033333   
5                               Downsview Central             0.033333   
6                             Downsview Northwest             0.033333   
7                                  Downsview West             0.033333   
8                Downsview,North Park,Upwood Park             0.033333   
9                                 Emery,Humberlea             0.033333   
10                     Fairview,Henry Farm,Oriole             0.033333   
11                Flemingdon Park,Don Mills South             0.033333   
12                                      Glencairn             0.033333   
13                              Hillcrest Village             0.033333   
14                                  Humber Summit             0.033333   
15                Lawrence Heights,Lawrence Manor             0.033333   
16                         Newtonbrook,Willowdale             0.033333   
17                 Northwood Park,York University             0.033333   
18                                      Parkwoods             0.033333   
19                        Silver Hills,York Mills             0.033333   
20                               Victoria Village             0.033333   
21                               Willowdale South             0.033333   
22                                Willowdale West             0.033333   
23                                York Mills West             0.033333   

    Burrito Place      Café  Coffee Shop  Fast Food Restaurant  Grocery Store  \
0        0.033333  0.066667          0.1              0.033333       0.033333   
1        0.033333  0.066667          0.1              0.033333       0.033333   
2        0.033333  0.066667          0.1              0.033333       0.033333   
3        0.033333  0.066667          0.1              0.033333       0.033333   
4        0.033333  0.066667          0.1              0.033333       0.033333   
5        0.033333  0.066667          0.1              0.033333       0.033333   
6        0.033333  0.066667          0.1              0.033333       0.033333   
7        0.033333  0.066667          0.1              0.033333       0.033333   
8        0.033333  0.066667          0.1              0.033333       0.033333   
9        0.033333  0.066667          0.1              0.033333       0.033333   
10       0.033333  0.066667          0.1              0.033333       0.033333   
11       0.033333  0.066667          0.1              0.033333       0.033333   
12       0.033333  0.066667          0.1              0.033333       0.033333   
13       0.033333  0.066667          0.1              0.033333       0.033333   
14       0.033333  0.066667          0.1              0.033333       0.033333   
15       0.033333  0.066667          0.1              0.033333       0.033333   
16       0.033333  0.066667          0.1              0.033333       0.033333   
17       0.033333  0.066667          0.1              0.033333       0.033333   
18       0.033333  0.066667          0.1              0.033333       0.033333   
19       0.033333  0.066667          0.1              0.033333       0.033333   
20       0.033333  0.066667          0.1              0.033333       0.033333   
21       0.033333  0.066667          0.1              0.033333       0.033333   
22       0.033333  0.066667          0.1              0.033333       0.033333   
23       0.033333  0.066667          0.1              0.033333       0.033333   

    Indonesian Restaurant  Japanese Restaurant  Juice Bar  Korean Restaurant  \
0                0.033333             0.033333   0.033333           0.066667   

In [60]:
# The new size is:
northyork_grouped.shape

(24, 24)

#### Below I uprint each neighborhood along with the top 3 most common venues.

In [61]:
num_top_venues = 3

for hood in northyork_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = northyork_grouped[northyork_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor,Downsview North,Wilson Heights----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Bayview Village----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Bedford Park,Lawrence Manor East----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----CFB Toronto,Downsview East----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Don Mills North----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Downsview Central----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Downsview Northwest----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Downsview West----
   

The following function sorts the venues in descending order.

In [62]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

I then create the new dataframe and display the top 5 venues for each neighbourhood.

In [64]:
num_top_venues = 7

indicators = ['st', 'nd', 'rd']

# I create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# I create a new dataframe
ny_neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
ny_neighbourhoods_venues_sorted['Neighbourhood'] = northyork_grouped['Neighbourhood']

for ind in np.arange(northyork_grouped.shape[0]):
        ny_neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(northyork_grouped.iloc[ind, :], num_top_venues)

ny_neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0  Bathurst Manor,Downsview North,Wilson Heights      Ramen Restaurant   
1                                Bayview Village      Ramen Restaurant   
2               Bedford Park,Lawrence Manor East      Ramen Restaurant   
3                     CFB Toronto,Downsview East      Ramen Restaurant   
4                                Don Mills North      Ramen Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop                  Café     Korean Restaurant   
1           Coffee Shop                  Café     Korean Restaurant   
2           Coffee Shop                  Café     Korean Restaurant   
3           Coffee Shop                  Café     Korean Restaurant   
4           Coffee Shop                  Café     Korean Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  
0  Vietnamese Restaurant         Movie Theater         Burrito Place  
1  Vietnamese Restaurant         Movie Theater         Burrito Place  
2  Vietnamese Restaurant         Movie Theater         Burrito Place  
3  Vietnamese Restaurant         Movie Theater         Burrito Place  
4  Vietnamese Restaurant         Movie Theater         Burrito Place

## 3. Clustering Neighbourhoods

I run k-means to cluster the neighborhood into 4 clusters.

In [66]:
# I set number of clusters
kclusters = 4

northyork_grouped_clustering = northyork_grouped.drop('Neighbourhood', 1)

# I run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(northyork_grouped_clustering)

# I then check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

I create a new dataframe that includes the cluster as well as the top 7 venues for each neighbourhood.

In [67]:
# add clustering labels
ny_neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

northyork_merged = northyork_data

# I merge northyork_merged with northyork_data to add latitude/longitude for each neighbourhood
northyork_merged = northyork_merged.join(ny_neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

northyork_merged.head() # to check the last columns!

Postcode     Borough               Neighbourhood   Latitude  Longitude  \
0      M2H  North York           Hillcrest Village  43.803762 -79.363452   
1      M2J  North York  Fairview,Henry Farm,Oriole  43.778517 -79.346556   
2      M2K  North York             Bayview Village  43.786947 -79.385975   
3      M2L  North York     Silver Hills,York Mills  43.757490 -79.374714   
4      M2M  North York      Newtonbrook,Willowdale  43.789053 -79.408493   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0      Ramen Restaurant           Coffee Shop   
1               0      Ramen Restaurant           Coffee Shop   
2               0      Ramen Restaurant           Coffee Shop   
3               0      Ramen Restaurant           Coffee Shop   
4               0      Ramen Restaurant           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                  Café     Korean Restaurant  Vietnamese Restaurant   
1                  Café     Korean Restaurant  Vietnamese Restaurant   
2                  Café     Korean Restaurant  Vietnamese Restaurant   
3                  Café     Korean Restaurant  Vietnamese Restaurant   
4                  Café     Korean Restaurant  Vietnamese Restaurant   

  6th Most Common Venue 7th Most Common Venue  
0         Movie Theater         Burrito Place  
1         Movie Theater         Burrito Place  
2         Movie Theater         Burrito Place  
3         Movie Theater         Burrito Place  
4         Movie Theater         Burrito Place

### Then we visualize the resulting clusters

In [69]:
# First I create map
map_northyork_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# I add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(northyork_merged['Latitude'], northyork_merged['Longitude'], northyork_merged['Neighbourhood'], northyork_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_northyork_clusters)
       
map_northyork_clusters

## 4. Examining Clusters

Now I examine each cluster and determine the discriminating venue categories that distinguish each cluster.

In [86]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 0, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Borough                                  Neighbourhood   Latitude  \
0   North York                              Hillcrest Village  43.803762   
1   North York                     Fairview,Henry Farm,Oriole  43.778517   
2   North York                                Bayview Village  43.786947   
3   North York                        Silver Hills,York Mills  43.757490   
4   North York                         Newtonbrook,Willowdale  43.789053   
5   North York                               Willowdale South  43.770120   
6   North York                                York Mills West  43.752758   
7   North York                                Willowdale West  43.782736   
8   North York                                      Parkwoods  43.753259   
9   North York                                Don Mills North  43.745906   
10  North York                Flemingdon Park,Don Mills South  43.725900   
11  North York  Bathurst Manor,Downsview North,Wilson Heights  43.754328   
12  North York                 Northwood Park,York University  43.767980   
13  North York                     CFB Toronto,Downsview East  43.737473   
14  North York                                 Downsview West  43.739015   
15  North York                              Downsview Central  43.728496   
16  North York                            Downsview Northwest  43.761631   
17  North York                               Victoria Village  43.725882   
18  North York               Bedford Park,Lawrence Manor East  43.733283   
19  North York                Lawrence Heights,Lawrence Manor  43.718518   
20  North York                                      Glencairn  43.709577   
21  North York               Downsview,North Park,Upwood Park  43.713756   
22  North York                                  Humber Summit  43.756303   
23  North York                                Emery,Humberlea  43.724766   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  -79.363452               0      Ramen Restaurant           Coffee Shop   
1  -79.346556               0      Ramen Restaurant           Coffee Shop   
2  -79.385975               0      Ramen Restaurant           Coffee Shop   
3  -79.374714               0      Ramen Restaurant           Coffee Shop   
4  -79.408493               0      Ramen Restaurant           Coffee Shop   
5  -79.408493               0      Ramen Restaurant           Coffee Shop   
6  -79.400049               0      Ramen Restaurant           Coffee Shop   
7  -79.442259               0      Ramen Restaurant           Coffee Shop   
8  -79.329656               0      Ramen Restaurant           Coffee Shop   
9  -79.352188               0      Ramen Restaurant           Coffee Shop   
10 -79.340923               0      Ramen Restaurant           Coffee Shop   
11 -79.442259               0      Ramen Restaurant           Coffee Shop   
12 -79.487262               0      Ramen Restaurant           Coffee Shop   
13 -79.464763               0      Ramen Restaurant           Coffee Shop   
14 -79.506944               0      Ramen Restaurant           Coffee Shop   
15 -79.495697               0      Ramen Restaurant           Coffee Shop   
16 -79.520999               0      Ramen Restaurant           Coffee Shop   
17 -79.315572               0      Ramen Restaurant           Coffee Shop   
18 -79.419750               0      Ramen Restaurant           Coffee Shop   
19 -79.464763               0      Ramen Restaurant           Coffee Shop   
20 -79.445073               0      Ramen Restaurant           Coffee Shop   
21 -79.490074               0      Ramen Restaurant           Coffee Shop   
22 -79.565963               0      Ramen Restaurant           Coffee Shop   
23 -79.532242               0      Ramen Restaurant           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                   Café     Korean Restaurant  Vietnamese Restaurant   
1                   Café     Korea

### Cluster 2

In [90]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 1, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Neighbourhood, Latitude, Longitude, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue]
Index: []

### Cluster 3

In [91]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 2, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Neighbourhood, Latitude, Longitude, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue]
Index: []

### Cluster 4

In [92]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 3, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Neighbourhood, Latitude, Longitude, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue]
Index: []

The result shows that ALL neighbourhoods have fallen into the same cluster, i.e., cluster 1. This is no wonder since they are in close proximity and have the venues in common.